types/wlr_input_method_v2.c

```diff
@@ -108,10 +116,126 @@ static void im_delete_surrounding_text(struct wl_client *client,
 	input_method->pending.delete.after_length = after_length;
 }
 
+bool wlr_surface_is_input_popup_surface_v2(struct wlr_surface *surface) {
+	return surface->role == &input_popup_surface_v2_role;
+}
+
+struct wlr_input_popup_surface_v2 *wlr_input_popup_surface_v2_from_wlr_surface(
+		struct wlr_surface *surface) {
+	assert(wlr_surface_is_input_popup_surface_v2(surface));
+	return (struct wlr_input_popup_surface_v2 *)surface->role_data;
+}
+
+void wlr_input_popup_surface_v2_send_text_input_rectangle(
+		struct wlr_input_popup_surface_v2 *popup_surface,
+		struct wlr_box *sbox) {
+	zwp_input_popup_surface_v2_send_text_input_rectangle(
+		popup_surface->resource, sbox->x, sbox->y, sbox->width, sbox->height);
```

Comment:
Protocol issue: the compositor has no way to know when the client has taken this new information into account. Thus the compositor might draw an old buffer at a new position, resulting in a glitch.